Implementation/Unit-Testing/index.md

```diff
@@ -4,39 +4,104 @@ versionFrom: 8.0.0
 
-## SetUp
+# Mocking
 
-The ```Current.Factory``` needs to be mocked before each unit test that has an Umbraco dependency, or you'll get an ```InvalidOperationException``` saying that  **"No factory has been set"**.
+When testing components in Umbraco, especially controllers, there are a few dependencies that needs to be mocked / faked in order to get your unit tests running. Every Umbraco controller has two constructors: one empty constructor without any parameters for anyone not interested in unit testing or dependency injections, and one with full constructor injection which contains all parameters needed for proper unit testing. A lot of these dependencies are interfaces, which are simply mocked using Mock.Of<>, but there are still a few explicit non-interface dependencies that needs to be faked. In this documentation all mocks and fakes have been placed in a base class to avoid having to repeat this setup in every test class.
 
 ```csharp
-[SetUp]
-public void SetUp()
+public abstract class UmbracoBaseTest 
 {
-    Current.Factory = new Mock<IFactory>().Object;
+    public ServiceContext ServiceContext;
+    public MembershipHelper MembershipHelper;
+    public UmbracoHelper UmbracoHelper;
+
+    public Mock<ICultureDictionary> CultureDictionary;
+    public Mock<ICultureDictionaryFactory> CultureDictionaryFactory;
+    public Mock<IPublishedContent> PublishedContent;
+    public Mock<IPublishedContentQuery> PublichedContentQuery;
+
+    [SetUp]
+    public virtual void SetUp()
+    {
+        this.SetupCultureDictionaries();
+        this.SetupPublishedContentQuerying();
+
+        this.ServiceContext = ServiceContext.CreatePartial();
+        this.MembershipHelper = new MembershipHelper(Mock.Of<HttpContextBase>(), Mock.Of<IPublishedMemberCache>(), Mock.Of<MembershipProvider>(), 
+        Mock.Of<RoleProvider>(), Mock.Of<IMemberService>(), Mock.Of<IMemberTypeService>(), Mock.Of<IUserService>(), Mock.Of<IPublicAccessService>(), AppCaches.NoCache, Mock.Of<ILogger>());
+        this.UmbracoHelper = new UmbracoHelper(Mock.Of<IPublishedContent>(), Mock.Of<ITagQuery>(), this.CultureDictionaryFactory.Object, 
+        Mock.Of<IUmbracoComponentRenderer>(), this.PublichedContentQuery.Object, this.MembershipHelper);
+    }
+
+    public virtual void SetupCultureDictionaries()
+    {
+        this.CultureDictionary = new Mock<ICultureDictionary>();
+        this.CultureDictionaryFactory = new Mock<ICultureDictionaryFactory>();
+        this.CultureDictionaryFactory.Setup(x => x.CreateDictionary()).Returns(this.CultureDictionary.Object);
+    }
+
+    public virtual void SetupPublishedContentQuerying()
+    {
+        this.PublishedContent = new Mock<IPublishedContent>();
+        this.PublichedContentQuery = new Mock<IPublishedContentQuery>();
+    }
+
+    public void SetupPropertyValue(string alias, object value, string culture = null, string segment = null)
```

Comment:
Would prob be better as a method? Generally you'll want to be testing isolated instances of IPublishedContent and not sharing the same one between each test. Thoughts?

Reply:
Makes sense. I updated the method so that it takes in IPublishedContent as a parameter, and removed the Ipublishedcontent mock from the base class to avoid sharing content between tests.